sbt-version-policy/src/main/scala/sbtversionpolicy/SbtVersionPolicySettings.scala

```diff
@@ -244,78 +247,133 @@ object SbtVersionPolicySettings {
+                    val affected = incompatibilityType match {
+                      case IncompatibilityType.BinaryIncompatibility => "current"
+                      case IncompatibilityType.SourceIncompatibility => "previous"
+                    }
+                    val howToFilter = problem.howToFilter.fold("")(hint => s"\n     filter with: ${hint}")
+                    s"   * ${problem.description(affected)}${howToFilter}"
+                  }.mkString("\n")
+                log.error(
+                  s"""Module ${formattedModule} is not ${formattedCompatibility} compatible with ${formattedPreviousModule}.
+                     |You have to relax our compatibility intention by changing the value of versionPolicyIntention, or to fix the incompatibilities.
+                     |We found the following incompatibilities:
+                     |${formattedProblems}""".stripMargin)
+                hadErrors = true
+              }
+            }
+            if (hadErrors) {
+              throw new MessageOnlyException("versionPolicyMimaCheck failed")
+            }
           }
         case None => Def.task {
-          // skip mima if no compatibility is intented
-          log.info(s"Not checking compatibility of module ${nameAndRevision(currentModule)} because versionPolicyIntention is set to 'Compatibility.None'")
+          // skip Mima if no compatibility is intended
+          log.info(s"Not checking compatibility of module ${formattedModule} because versionPolicyIntention is set to 'Compatibility.None'")
         }
       }
-    }.value
+    }.value,
+    versionPolicyFindIssues := Def.ifS((versionPolicyFindIssues / skip).toTask)(Def.task {
+      streams.value.log.debug("Not finding incompatibilities with previous releases because 'versionPolicyFindIssues / skip' is 'true'")
+      Seq.empty[(ModuleID, (DependencyCheckReport, Seq[(IncompatibilityType, Problem)]))]
+    })(
+      Def.ifS[Seq[(ModuleID, (DependencyCheckReport, Seq[(IncompatibilityType, Problem)]))]](Def.task {
+        versionPolicyPreviousVersions.value.isEmpty
+      })(Def.task {
+        throw new MessageOnlyException("Unable to find compatibility issues because versionPolicyPreviousVersions is empty.")
+      })(Def.task {
+        versionPolicyPreviousVersions.value
+        val dependencyIssues = versionPolicyFindDependencyIssues.value
+        val mimaIssues = versionPolicyFindMimaIssues.value
+        assert(
+          dependencyIssues.map(_._1.revision).toSet == mimaIssues.map(_._1.revision).toSet,
+          "Dependency issues and Mima issues must be checked against the same previous releases"
+        )
+        for ((previousModule, dependencyReport) <- dependencyIssues) yield {
+          val (_, problems) =
+            mimaIssues
+              .find { case (id, _) => previousModule.revision == id.revision }
+              .get // See assertion above
+          previousModule -> (dependencyReport, problems)
+        }
```

Comment:
I'm having a bit of trouble understanding this code, I think because I'm not sure exactly what range of values it's operating over. `dependencyIssues` & `mimaIssues` are both `Seq[(sbt.ModuleID, ...]` but what range of `ModuleID` do we expect that to be? I assume that `ModuleID` means a subproject within the sbt project we are running in, but I'm not sure of:

* will all `ModuleId`s be for the same module, but potentially for many different revisions? Would it be reasonable to assert that we're only expecting _one_ common revision?
* if `ModuleId`s can be for differing modules (eg `core` & `extra-thing`), then linking them together just by revision (`previousModule.revision == id.revision`) wouldn't work, right?

Reply:
That’s a great question. The module IDs we iterate over are the previous versions of the same module that we check the compatibility against.

It is possible to check the compatibility against several previous versions, not just one, although by default only the previous release is tested. This can be customized by the users via the key `versionPolicyPreviousVersions`.

We inherited this behavior because Mima itself works like that, although I am not sure it is that important that we follow the same approach in sbt-version-policy. I decided not to change that in this PR.

So, ideally we’d have a range of revisions, not a range of module IDs, but `versionPolicyFindDependencyIssues` already used a range of module IDs so I followed the same structure for `versionPolicyFindMimaIssues` as well (and also `versionPolicyFindIssues` and `versionPolicyAssessCompatibility`…).

Reply:
Ah, thanks for explaining that! It does make me wonder whether `versionPolicyAssessCompatibility`, being a new key, needs to return `Seq[(ModuleID, Compatibility)]`, or whether it could be just `(ModuleID, Compatibility)`, giving info for the latest published revision. Anything that makes the output simpler to model and easier to parse is good from my perspective - would it be harmful to sbt-version-policy to make that simplification on this new key?

Reply:
I see two arguments to keep the `Seq`:

- it allows the users to compute the compatibility against multiple previous versions if they want to (although I have no clue whether people really need that…), without adding much costs on our side.
- we currently implement the `skip` behavior by returning an empty collection. Alternatively, we could use an `Option`. A better solution would be to “assign” the task to its definition (the part where we do `versionPolicyAssessCompatibility := { … }`) only if `skip` is `false`, but I am not sure sbt supports that.